src/nvim/getchar.c

```diff
@@ -237,6 +237,57 @@ char_u *get_inserted(void)
   return get_buffcont(&redobuff, FALSE);
 }
 
+///
+/// Prepend string to the head of current block of the given buffer.
+///
+static void prepend_buf(buffheader_T *buf, char_u *s)
+{
+  size_t slen = STRLEN(s);
+
+  if (buf->bh_index >= slen) {
+    // Easy case
+    buf->bh_index -= slen;
+    memmove(buf->bh_first.b_next->b_str + buf->bh_index, s, slen);
+  } else {
+    // Need to allocate a new buffer block
+    size_t len;
+    if (slen < MINIMAL_SIZE) {
+      len = MINIMAL_SIZE;
+    } else {
+      len = (size_t)slen;
+    }
+    buffblock_T *p = xmalloc(sizeof(buffblock_T) + len);
+    memmove(p->b_str, s, slen);
+    char_u *str_head = buf->bh_first.b_next->b_str + buf->bh_index;
+    memmove(p->b_str + slen, str_head, STRLEN(str_head) + 1);
```

Comment:
Looks like this procedure came from `ins_typebuf()`.

Reply:
Yes, it is a kind of duplicate of a part of `ins_typebuf()`, however `ins_typebuf()` is only for inserting into `typebuf`, so we need this for read buffers.